src/access/acp.ts

```diff
@@ -20,24 +20,48 @@
  */
 
 import { WithAccessibleAcr } from "../acp/acp";
-import { getPolicyUrlAll } from "../acp/control";
+import { getAcrPolicyUrlAll, getPolicyUrlAll } from "../acp/control";
 import { internal_getAcr } from "../acp/control.internal";
+import { getPolicyAll } from "../acp/policy";
 import { getRuleAll } from "../acp/rule";
+import { acp } from "../constants";
 import { WithResourceInfo } from "../interfaces";
 import { getSourceIri } from "../resource/resource";
-import { asUrl } from "../thing/thing";
+import { getSolidDataset } from "../resource/solidDataset";
+import { getUrlAll } from "../thing/get";
+import { asUrl, getThing } from "../thing/thing";
 
 export function internal_hasInaccessiblePolicies(
   resource: WithAccessibleAcr & WithResourceInfo
 ): boolean {
   const sourceIri = getSourceIri(resource);
-  const policyUrls = getPolicyUrlAll(resource);
-  const ruleUrls = getRuleAll(internal_getAcr(resource)).map((rule) =>
-    asUrl(rule)
-  );
-  // If either a policy or a rule are not defined in the ACR, return false
+
+  // Collect all policies that apply to the resource or its ACR (aka active)
+  const activePolicyUrls = getPolicyUrlAll(resource);
+  getAcrPolicyUrlAll(resource).forEach((policyUrl) => {
+    activePolicyUrls.push(policyUrl);
+  });
+
+  // Collect all the rules referenced by the active policies.
+  const ruleUrls: string[] = [];
+  activePolicyUrls.forEach((policyUrl) => {
+    const acr = internal_getAcr(resource);
+    const policyThing = getThing(acr, policyUrl);
+    if (policyThing !== null) {
+      getUrlAll(policyThing, acp.anyOf).forEach((activeRuleUrl) =>
+        ruleUrls.push(activeRuleUrl)
+      );
+      getUrlAll(policyThing, acp.allOf).forEach((activeRuleUrl) =>
+        ruleUrls.push(activeRuleUrl)
+      );
+      getUrlAll(policyThing, acp.noneOf).forEach((activeRuleUrl) =>
+        ruleUrls.push(activeRuleUrl)
+      );
+    }
+  });
```

Comment:
Again a matter of style (so feel free to ignore!), and `.flat()` is not supported in IE so would need addition to the list of polyfills mentioned in the README, but possibly interesting:

```suggestion
  const ruleUrls: string[] = activePolicyUrls.map(policyUrl => {
    const acr = internal_getAcr(resource);
    const policyThing = getThing(acr, policyUrl);
    if (policyThing === null) {
      return [];
    }
    return getUrlAll(policyThing, acp.anyOf)
      .concat(getUrlAll(policyThing, acp.allOf))
      .concat(getUrlAll(policyThing, acp.noneOf))
  }).flat();
```

Reply:
Again, very good suggestion, thanks ! It does look more readable and less convoluted. 4646c9c9066b76a173b5c9591a46485bd7bdf3dd

Reply:
Well nevermind, `.flat` isn't supported in Node10 ^^

Reply:
Blessed be small commits 4b8093de4910785f3f321665c3d4a72d1a291b45

Reply:
Aargh. Just four more months... :)